contracts/Witch.sol

```diff
@@ -69,138 +75,205 @@ contract Witch is AccessControl() {
+        require(limits_.sum <= limits_.line * (10**limits_.dec), "Collateral limit reached");
 
         limits[vault_.ilkId] = limits_;
-        auctions[vaultId] = Auction({
-            owner: vault_.owner,
-            start: block.timestamp.u32()
-        });
+        auctions[vaultId] = Auction({owner: vault_.owner, start: block.timestamp.u32()});
         cauldron.give(vaultId, address(this));
         emit Auctioned(vaultId, block.timestamp.u32());
     }
 
+    /// @dev Private fn called by buy() and payAll()
+    /// @param ilkId Id of asset used for collateral
+    /// @param auctionStart Block timestamp when auction was started
+    /// @param duration Time that auctions take to go to minimal
+    /// @param p Initial Offer. Proportion of collateral sold at auction start(1e18 = 100%)
+    /// @param artIn Portion of debt being bought (in terms of base)
+    /// @param totalArt Total debt
+    /// @param totalInk Total collateral
+    /// @param min Minimum amount of collateral acceptable by buyer
+    /// @return inkOut Amount of collateral
+    function _inkOut(
+        bytes6 ilkId,
+        uint32 auctionStart,
+        uint32 duration,
+        uint64 p,
+        uint256 artIn,
+        uint128 totalArt,
+        uint128 totalInk,
+        uint128 min
+    ) private returns (uint256 inkOut) {
+        Limits memory limits_ = limits[ilkId];
+
+        // If the world has not turned to ashes and darkness, auctions will malfunction on
+        // the 7th of February 2106, at 06:28:16 GMT
+        // TODO: Replace this contract before then 😰
+        // UPDATE: Added reminder to Google calendar ✅
+        uint256 elapsed;
+        unchecked {
+            elapsed = uint32(block.timestamp) - auctionStart;
+        }
+
+        uint256 t = elapsed > duration ? 1e18 : elapsed.wdivup(duration);
+
+        //          (      a         )
+        // inkOut = (artIn / totalArt) * totalInk * (p + (1 - p) * t)
+        {
+            uint256 a = uint256(artIn).wdivup(totalArt);
+            inkOut = a.wmul(totalInk).wmulup(uint256(p) + uint256(1e18 - p).wmulup(t));
+        }
+
+        require(inkOut >= min, "Not enough bought");
+        require(totalArt == artIn || totalInk - inkOut >= limits_.dust * (10**limits_.dec), "Leaves dust");
+        limits[ilkId].sum = limits_.sum - inkOut.u128();
+    }
+
     /// @dev Pay `base` of the debt in a vault in liquidation, getting at least `min` collateral.
     /// Use `payAll` to pay all the debt, using `buy` for amounts close to the whole vault might revert.
-    function buy(bytes12 vaultId, uint128 base, uint128 min)
-        external
-        returns (uint256 ink)
-    {
-        require (auctions[vaultId].start > 0, "Vault not under auction");
+    /// @param vaultId Id of vault to buy
+    /// @param base Amount of base to pay
+    /// @param min Minimum amount of collateral that must be received
+    /// @return ink Amount of vault collateral sold
+    function buy(
```

Comment:
You can do an internal `_buy` function that takes the vault and balances structs as parameters, along with whatever else is needed. Then `paySome` and `payAll` would be thin functions that just read a couple of structs and call the single `_buy` function that has the logic.